packages/datastore/src/listfield.ts

```diff
@@ -168,6 +168,36 @@ class ListField<T extends ReadonlyJSONValue> extends Field<ListField.Value<T>, L
     return { value: clone, change };
   }
 
+  /**
+   * Encode a system patch so that it can be sent across a network.
+   *
+   * @param patch - The patch to encode.
+   *
+   * @returns a JSON value that can be sent across the network.
+   */
+  encodePatch(patch: ListField.Patch<T>): ListField.Patch<T> {
+    return patch.map(part => ({
+      ...part,
+      removedIds: part.removedIds.map(id => encodeId(id)),
+      insertedIds: part.insertedIds.map(id => encodeId(id))
+    }));
```

Comment:
Would it be more type-safe to use `ListField.mergePatch()`?

Reply:
Hmm, I guess it won't actually, since that would require a full `Patch` object...

Reply:
I think the problem here was that `Patch` derives from `ReadonlyJSONObject`, so attributes that haven't been previously defined are, in fact, allowed.

@vidartf you have recently done a deep dive on the type-safety of JSON objects. Do you have any ideas about how it could be tightened up? Maybe by not deriving from `ReadonlyJSONObject` in the first place?

Reply:
@ian-r-rose I don't think `Patch` derives from `ReadonlyJSONObject` ?

Reply:
(although it takes a generic arg that extends `ReadonlyJSONObject`)

Reply:
Note that this code fails:

```ts
  encodePatch(patch: ListField.Patch<T>): ListField.Patch<T> {
    return patch.map(part => {
      const ret: ListField.PatchPart<T> = {
        ...part,
        removeIds: part.removedIds.map(id => id),
        inserteIds: part.insertedIds.map(id => id)
      }
      return ret;
    });
  }
```

Reply:
Adding a function like this in `Private` also picks up the error:
```ts

  export
  function updatePatchPart<T>(part: ListField.PatchPart<T>, update: Partial<ListField.PatchPart<T>>): ListField.PatchPart<T> {
    return {
      ...part,
      ...update
    }
  }
```

Reply:
You are right that `ListField.Patch` doesn't explicitly extend `ReadonlyJSONValue`, however the interface that it satisfies [does](https://github.com/jupyterlab/lumino/blob/a1fe9493ec962ba371e8c05f68ae3cce089f4ee1/packages/datastore/src/field.ts#L19).

Your `PatchPart` fix does indeed catch the error; I'd like to have a good understanding of why a list of `PatchPart`s does not, though.

Reply:
> I'd like to have a good understanding of why a list of PatchParts does not, though.

I think the original code had too many degrees of separation between the attributes being specified, and the outer type being specified. I.e. I don't think the return type of the outer function is propagated deep enough in to test the object declaration. That's my working hypothesis at least :)